### DISCUSSION TITLE
Scattering coefficient

### DISCUSSION THREAD

**Discussion by liuxoB** [NVlabs/sionna#188] [Q&A] [answered]
Dear developers and others, 

I want to do simulations with scattering and I need to specify the `scattering_coefficient` for the radio materials. However, I want that the signal power of every ray radiating from the surface is proportional to the `scattering_pattern` and not an additional coefficient. 

If I have a simple simulation with 1 flat surface the signal reflects from and the simulation yields 100 paths of which 1 is the reflection and the rest is scattered light, then I need to put the `scattering_coefficient` very small to make all rays equal in power (disregarding the `scattering_pattern`). But if I then do a simulation with a much smaller surface of which I get 5 rays (1 refl, 4 scat) from the simulation, then I need to set the `scattering_coefficient` much larger to get a similar result. 

Is there a way to bypass this problem to make sure the power distribution between rays is identical no matter how many rays are reflected from a surface? I was thinking maybe I can trick sionna to think every reflected path is simply a scattered path and then put the `scattering_coefficient` to 1, but I could not figure out a way to make that work.

best, Floris

**Comment by jhoydis** [comment]
Hi Floris,

I am not sure to fully understand your issue.
Could you provide a code example showing what you are trying to accomplish? 

In general, the larger the reflecting surface, the larger the number of scattered rays that you see. The energy of the specularly reflected path does not change with the size of the surface. The energy per scattered ray should be more or less constant.

**Reply by liuxoB** [reply]
Thank you for your answer! There is not much code to go with this problem. I am just tweaking the surface size and scattering coefficient and then check what the attenuation factors of the paths look like. It always seems that the attenuation factor of the specularly reflected path is much different from all other paths, even if the paths nearly follow the same trajectory. 

Here is some sample code where I vary the amount of rays kept (instead of increasing the surface size.) What I want is for the specular reflection to be in the same order of magnitude as all other reflection (this makes sense to me since they are all nearly the same angle). 
If I decrease the scat_keep_prob (lowest part of the code) then the power of the scattered rays will increase proportionally to make sure the power output is still the same. As can be seen is that the relative power of the specular reflected ray is now larger than any other ray. I would like this power difference to be in line with the scattering pattern, no matter how many rays I choose or how large the surface is.

```
import sionna
import tensorflow as tf
import numpy as np
scene = sionna.rt.load_scene(filename=sionna.rt.scene.simple_reflector)
for obj in scene.objects:
    refl = scene.get(obj)
    refl.radio_material.scattering_coefficient=0.1

scene.tx_array = sionna.rt.PlanarArray(num_rows=1, num_cols=1, 
                                       vertical_spacing=0.5,
                                       horizontal_spacing=0.5, pattern="iso",
                                       polarization="V")
scene.rx_array = scene.tx_array



# transceiver
position = [0,0,50]
scene.add(sionna.rt.Transmitter(name="tx", position=position))
scene.add(sionna.rt.Receiver(name="rx", position=position)) #[x,y,height]

tf.random.set_seed(1) 
paths = scene.compute_paths(max_depth=3, 
                            los=False, 
                            num_samples=1e6, 
                            scattering=True, 
                            scat_keep_prob=0.5)

a, tau = paths.cir(los=False,          # Do not consider Line of Sight paths.
                   reflection=True, 
                   diffraction=True, 
                   scattering=True) 

print("path types:", paths.types)
print("absolute attenuation factors:", abs(a))
print("Power:", tf.reduce_sum(abs(a)**2))
scene.preview(paths=paths)
```
path types: tf.Tensor([[1 1 3 3 3 3 3 3 3 3 3 3 3 3 3 3]], shape=(1, 16), dtype=int32)
absolute attenuation factors: tf.Tensor(
[[[[[[[0.0000000e+00]
      [6.7807021e-05]
      [3.8550681e-08]
      [3.8550336e-08]
      [3.8550223e-08]
      [3.8549867e-08]
      [3.8548954e-08]
      [3.8548830e-08]
      [3.8548713e-08]
      [3.8548258e-08]
      [3.8548137e-08]
      [3.8548013e-08]
      [3.8547913e-08]
      [3.8547672e-08]
      [3.8547562e-08]
      [3.8546638e-08]]]]]]], shape=(1, 1, 1, 1, 1, 16, 1), dtype=float32)
Power: tf.Tensor(4.597812e-09, shape=(), dtype=float32)

Next cell: 
```
tf.random.set_seed(1) 
paths = scene.compute_paths(max_depth=3, 
                            los=False, 
                            num_samples=1e6, 
                            scattering=True, 
                            scat_keep_prob=0.1)

a, tau = paths.cir(los=False,          # Do not consider Line of Sight paths.
                   reflection=True, 
                   diffraction=True, 
                   scattering=True) 

print("path types:", paths.types)
print("absolute attenuation factors:", abs(a))
print("Power:", tf.reduce_sum(abs(a)**2))
scene.preview(paths=paths)
```
path types: tf.Tensor([[1 1 3 3 3]], shape=(1, 5), dtype=int32)
absolute attenuation factors: tf.Tensor(
[[[[[[[0.000000e+00]
      [6.780702e-05]
      [8.619809e-08]
      [8.619652e-08]
      [8.619625e-08]]]]]]], shape=(1, 1, 1, 1, 1, 5, 1), dtype=float32)
Power: tf.Tensor(4.5978146e-09, shape=(), dtype=float32)

**Reply by jhoydis** [reply, marked as answer]
Hi Floris,

There is no physical reason why the diffusely reflected rays should carry the same amount of energy as the reflected ray.
The biggest difference between both is that the energy of the specular path decays as $d^{-2}$, where $d$ is the total distance, while that of a diffusely reflected path decays as $(d_1 d_2)^{-2}$, where $d_1$ is the distance to the scattering object and $d_2$ is the distance from the latter to the receiver.

Please have a look at the [EM Primer](https://nvlabs.github.io/sionna/em_primer.html#scattering) which explains the different propagation phenomena.

In the far-field, the sum of the energy of the scattered rays from a surface towards a specific point is constant and depends only on the surface area, the scattering coefficient and scattering pattern, the constituent parameters ($\varepsilon_r, \sigma$), as well as the incoming and outgoing angle. You can approximate the scattered field by a single scattered ray carrying all of the power or multiple rays which carry each only a part of the energy. This is shown in the [Scattering tutorial](https://nvlabs.github.io/sionna/examples/Sionna_Ray_Tracing_Scattering.html#Validation-Against-the-“Far”-Wall-Approximation).

You can think about scattering in the following way: Cut you surface into many small tiles. Each tile captures some energy which is proportional to its area and the cosine of the incoming angle with the surface normal. A part of this energy is now reradiated as if the surface tile was a point source with the scattering pattern as antenna pattern. In the limit, you integrate over infinitesimally small surface elements. Now, when you change `scat_keep_prob` or `num_samples` in the code above, you implicitly change the size of the surface tiles and hence the number of scattered rays. (Note that in Sionna RT, surfaces are not explicitly tiled. This happens implicitly through the rays which cover approximately equal solid angles $\Delta\Omega$. When a ray hits a surface, it illuminates an area $\Delta A=\Delta\Omega d^2 \cos(\theta)$, where $d$ is the travelled distance and $\theta$ the incoming angle with the surface normal. This area will be used to compute the energy of the corresponding scattered ray.) 

How many scattered rays you want to simulate depends on the available memory you have. The parameter `scat_keep_prob` allows you to avoid dealing with too many scattered rays.

**Reply by liuxoB** [reply]
Thank you for the detailed answer! I somewhat understand now. The only thing I am now still struggling with is what value to choose for the scattering_coefficient. If I have a very far away flat surface that perfectly diffusely scatters light (Lambertian). What would be a suitable scattering coefficient for that? If I put this value to 1, and I happen to get a single specular reflected ray to the receiver (because the object is so far and the surface normal points to the transceiver), then I would receive no return signal, but intuitively there should be some signal returned. Does that mean I am simulating with too little amount of rays? 

Or in the same scenario: if I put the scattering coefficient as 0.5, and it only finds the specular path, then the intensity of the signal received will be smaller than it should be, since there would have been light scattered back but it wasnt found. 

Does this basically mean that for a realistic simulation, it would need to find at least 1 scattered ray? And where do I even find realistic values for the scattering coefficient for the itu materials for example?

**Reply by jhoydis** [reply]
The scattering coefficient determines how much energy of the specularly reflected ray is taken away from it and used for diffuse reflection. If it is equal to one, the specular path has zero energy. 

In general, the scattering coefficient is a material parameter which is related to the roughness. Please have a look at the [references](https://nvlabs.github.io/sionna/examples/Sionna_Ray_Tracing_Scattering.html#References).
Also note that the scattering patterns have various parameters that can be tuned.
There are to my best knowledge no databases with scattering parameters available that can be used, similar to the ITU materials.

The directive model with $\alpha_r=10$ and a scattering coefficient in the range of 0.2-0.4 gave some good agreement with measurements for a specific scenario. However, this really depends on the scene and carrier frequency (in general, the higher the frequency, the rougher surfaces appear).

If you just want to try something, I would probably go with Lambertian scattering and a scattering coefficient of 0.4. But again, if this makes sense or not depends on your specific scenario.

The goal of differentiable ray tracing is that one can learn such material parameters from measurements, i.e., the channel coefficients
 are differentiable with respect to the scattering coefficient.
